### PR TITLE
Defer loading of videos on details page

### LINKED ISSUE #3285: Defer loaded of video on details page to prevent render blocking

### PR DESCRIPTION
**Pull request by steverydz**
## Done

Defer loading of video on snap details page so it isn't render blocking

## Issue / Card

Fixes #3285

## QA

- Go to https://snapcraft-io-3286.demos.haus/code
- Go to https://snapcraft-io-3286.demos.haus/tbmb-bsi-test
- In both cases check that the video loads and the content doesn't jump

### CODE REVIEW

File: static/sass/_snapcraft_p-video-wrapper.scss [hunk 1]
```diff
@@ -0,0 +1,7 @@
+@mixin snapcraft-p-video-wrapper {
+  .p-video-wrapper {
```
Comment (anthonydillon) [inline static/sass/_snapcraft_p-video-wrapper.scss:2]: Can you use the Vanilla one? https://vanillaframework.io/docs/utilities/embedded-media

**Review by tbille — APPROVED** [review]
I :+1: this

**Review by anthonydillon — CHANGES_REQUESTED** [review]
Comment on the new utility.

Also, I true believer auto playing promotional media is not required and should be discouraged. I don't know of any other store that auto plays videos on details pages. Who is requesting this?

**Comment by tbille** [comment]
@anthonydillon the autoplay has been already a long debate a while ago. I can ask to get who asked for it.

**Comment by steverydz** [comment]
@anthonydillon I've swapped it out for the vanilla pattern and disabled autoplay on the yt videos

File: templates/partials/_video.html [hunk 2]
```diff
@@ -1,10 +1,10 @@
 {% if video.type == "youtube" %}
   <iframe id="ytplayer" type="text/html" width="818" height="460"
-          src="{{ video.url }}?autoplay=1&mute=1&cc_load_policy=1&modestbranding=1&rel=0" frameborder="0"></iframe>
+          src="" frameborder="0" data-src="{{ video.url }}?autoplay=0&mute=1&cc_load_policy=1&modestbranding=1&rel=0"></iframe>
```
Comment (anthonydillon) [inline templates/partials/_video.html:3]: Now that the video is not autoplay you can you remove the mute option.

File: templates/partials/_snap-details-video-layout.html [hunk 3]
```diff
@@ -17,3 +17,18 @@
     </div>
   {% endif %}
 </div>
+
+<script>
+  document.addEventListener("DOMContentLoaded", function() {
+    const youtubeVideoFrame = document.getElementById("ytplayer");
+    const vimeoplayerFrame = document.getElementById("vimeoplayer");
```
Comment (anthonydillon) [inline templates/partials/_snap-details-video-layout.html:24]: The videos are embedded in a loop. Therefore, I assume there could be multiple with the same `id`?

File: templates/partials/_snap-details-video-layout.html [hunk 4]
```diff
@@ -4,7 +4,7 @@
    see https://github.com/canonical-web-and-design/vanilla-framework/issues/2448 #}
   <div class="col-10 u-align--center">
     {% for video in videos %}
-      <div class="js-video-slide" data-video-type="{{video.type}}" data-video-url="{{video.url}}" data-video-id="{{video.id}}">
+      <div class="js-video-slide u-embedded-media" data-video-type="{{video.type}}" data-video-url="{{video.url}}" data-video-id="{{video.id}}">
```
Comment (anthonydillon) [inline templates/partials/_snap-details-video-layout.html:7]: The videos are embedded in a loop. Therefore, I assume there could be multiple with the same `id`?

File: templates/partials/_snap-details-video-layout.html [hunk 5]
```diff
@@ -17,3 +17,18 @@
     </div>
   {% endif %}
 </div>
+
+<script>
+  document.addEventListener("DOMContentLoaded", function() {
```
Comment (anthonydillon) [inline templates/partials/_snap-details-video-layout.html:22]: Do you think this script would be better in the `_video.html` particle as its closer to the elements it's interacting with?

**Review by anthonydillon — CHANGES_REQUESTED** [review]
Thanks Steve, few more questions in the code

File: templates/partials/_video.html (hunk 2 above)
Comment (steverydz) [reply templates/partials/_video.html:3]: Done

File: templates/partials/_snap-details-video-layout.html (hunk 3 above)
Comment (steverydz) [reply templates/partials/_snap-details-video-layout.html]: There can only be one video. I've refactored the python to reflect this.

File: templates/partials/_snap-details-video-layout.html (hunk 4 above)
Comment (steverydz) [reply templates/partials/_snap-details-video-layout.html:6]: same as above

File: templates/partials/_snap-details-video-layout.html (hunk 5 above)
Comment (steverydz) [reply templates/partials/_snap-details-video-layout.html]: Moved

**Comment by steverydz** [comment]
@anthonydillon I've addressed all of your comments

File: templates/partials/_video.html (hunk 2 above)
Comment (anthonydillon) [reply templates/partials/_video.html:3]: Sorry I meant you could remove the mute param altogether. It is now muted when you play.

Comment (steverydz) [reply templates/partials/_video.html]: ah, I did remove it but accidentally reverted when fixing the python linting

File: webapp/store/logic.py [hunk 6]
```diff
@@ -427,10 +425,13 @@ def get_icon(media):
...
 
-def get_videos(media):
-    return [
-        get_video_embed_code(m["url"]) for m in media if m["type"] == "video"
-    ]
+def get_video(media):
+    video = None
+    for m in media:
+        if m["type"] == "video":
+            video = get_video_embed_code(m["url"])
```
Comment (anthonydillon) [inline webapp/store/logic.py:432]: Could you just return here which means you don't need the `video` variable or the break and return below?

**Review by anthonydillon — APPROVED** [review]
LGTM now with one suggestion on the code

**Comment by steverydz** [comment]
> LGTM now with one suggestion on the code

For that to work wouldn't we need to wrap the whole return value in `get_video_embed_code` which would make it pretty long and hard to reason about?

File: webapp/store/logic.py [hunk 7]
```diff
@@ -427,10 +425,13 @@ def get_icon(media):
...
+            break
+    return video
```
Comment (anthonydillon) [inline webapp/store/logic.py:434]: Mmm ... I don't think so. I'm suggesting:
```python
def get_video(media):
    video = None
    for m in media:
        if m["type"] == "video":
            return get_video_embed_code(m["URL"])
```

**Pull request merged**